app/assets/javascripts/tylium/modules/auto_save/local.js

```diff
@@ -0,0 +1,126 @@
+    this.restoreData();
+  }
+
+  behaviors() {
+    var that = this;
+
+    this.target.addEventListener('submit', function(event) {
+      that.submitted = true;
+      localStorage.removeItem(that.key);
+    });
+
+    var clearCacheElement = this.target.querySelector('[data-behavior~=clear-local-auto-save]');
+
+    if (clearCacheElement) {
+      clearCacheElement.addEventListener('click', function(event) {
+        that.cancelled = true;
+        localStorage.removeItem(that.key);
+      })
+    }
+
+    this.target.querySelectorAll('input, textarea, select').forEach(function(input) {
+      // Don't add event handler for submit button
+      if (input.type === 'button') { return true; }
+
+      // we're using a jQuery plugin for :textchange event, so need to use $()
+      $(input).on('textchange change', that.handleTextChange.bind(that));
+    })
+  }
+
+  handleTextChange() {
+    clearTimeout(this.debounceTimeout);
+
+    this.debounceTimeout = setTimeout(function() {
+      this.setData();
+    }.bind(this), this.debounceTimer);
+  }
+
+  getData() {
+    var that = this;
+    var data = new FormData(this.target);
+    var dataArray = Array.from(data).filter(function(dataElement) {
+      return !that.excludedInputNames.includes(dataElement[0] );
+    });
+
+    return dataArray;
+  }
+
+  restoreData() {
+    var that = this;
+    var data = JSON.parse(localStorage.getItem(this.key));
+
+    if (data !== null) {
+      // Deserialization of form data
+      // Reference: https://wildwolf.name/store-formdata-object-in-localstorage/
+      for (var item of data) {
+        var name = item[0];
+        var element = that.target.elements[name];
+
+        // Skip loop if element is undefined
+        if (!element) { continue; }
+
+        // RadioNodeList is an array of checkboxes or radio buttons
+        if (element instanceof RadioNodeList) {
+          element.forEach(function(radioNodeListElement) {
+            if (['checkbox', 'radio'].includes(radioNodeListElement.type)) {
+              if (radioNodeListElement.value === item[1]) {
+                radioNodeListElement.checked = true;
+              }
+            }
+          })
+        } else if (element.type === 'checkbox') {
+          if (element.value === item[1]) {
+            element.checked = true;
+          }
+        } else if (element.type === 'file') {
+          element.value = '';
+        } else {
+          element.value = item[1];
+        }
+
+        $(element).trigger('load-preview');
+        this.restoreTagInputDisplay(data);
+      }
+    }
+  }
+
+  restoreTagInputDisplay(data) {
+    var tagInputData = data.find(function(item) {
+      return item[0] === 'issue[tag_list]';
+    })
+
+    if (tagInputData) {
+      var hiddenInput = document.querySelector('#issue_tag_list');
+      hiddenInput.value = tagInputData[1];
+
+      var $target = $(`.js-taglink[data-tag='${tagInputData[1]}']`);
+      new SelectTagDropdown($target)
+    }
+  }
```

Comment:
There's still something here that I feel could be improved.

We don't have to, but me may _want_ to.

First L105-L107 are repeating a loop that we already did in L72. Can't we just add the `if issue[tag_list]' conditional there, and only then make the call to `SelectTagDrop`?

Second, if we're adding a separate class whose responsibility is to "select the tag", then `LocalAutoSave` has no business touching the `#issue_tag_list`, that responsibility is a lot closer to the component that "selects the tag" than to the auto-save component. Auto-save cares about persisting the data, but if we have another object whose responsibility is to manage the state of the Tag select control, let it do it.

Particularly L111 is also in `SelectTagDropdown:14`, which reinforces my thinking that we're splitting/duplicating reponsibilities.

Right now the `SelectTagDropdown` class only does some cosmetics (CSS, label, etc.) I'd say we move L14 up to the `SelectTagDropdown` class, removed it from `LocalAutoSave` and let `SelectTagDropdown` handle the hidden input value and the CSS of the select control.

Reply:
Thanks @etdsoft , I've refactored it to be inside the restoreData() loop and it's much cleaner now!
```javascript
if (element.name === 'issue[tag_list]') {
  var $tagDropdownItem = $(`.js-taglink[data-tag='${item[1]}']`);
  new SelectTagDropdown($tagDropdownItem)
}
```
Initially, I tried to trigger a onclick event programmatically on the $tagDropdownItem but it doesn't seem to work on a hidden element, so i went with the above instead